rules/store.go

```diff
@@ -0,0 +1,191 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE
+
+package rules
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/m3db/m3cluster/kv"
+	"github.com/m3db/m3metrics/generated/proto/schema"
+)
+
+var (
+	errNilRuleSet    = errors.New("nil ruleset")
+	errNilNamespaces = errors.New("nil namespaces")
+)
+
+// Store facilitates read/write operations to the backing kv store.
+type Store interface {
+	// WriteRuleSet saves the given ruleset to the backing store.
+	WriteRuleSet(MutableRuleSet) error
+
+	// WriteAll saves both the given rulset and namespace to the backing store.
+	WriteAll(*Namespaces, MutableRuleSet) error
+
+	// ReadNamespaces returns the version and the persisted namespaces in kv store.
+	ReadNamespaces() (*Namespaces, error)
+
+	// ReadRuleSet returns the version and the persisted ruleset data in kv store.
+	ReadRuleSet(namespaceName string) (MutableRuleSet, error)
```

Comment:
I think that's actually a good practice to get immutable objects back and if the caller wants to mutate it, they call `.ToMutableRuleSet()` to be explicit about their purpose.